handle data-link-id in prepareToReset

cathysax opened the pull request.

cathysax committed:
handle data-link-id in prepareToReset

tabascq commented on puzzlejs/puzzle.js:
```diff
@@ -1462,11 +1462,15 @@ function PuzzleGrid(puzzleEntry, options, container, puzzleId) {
+            var linkId = s.getAttribute("data-link-id");
+            var asel = []
+            if (extractId) asel.push("table:not(.copy-only) .extract[data-extract-id='" + extractId + "']");
+            if (linkId) asel.push("table:not(.copy-only) .cell[data-link-id='" + linkId + "']");
```
this can be `.link`, not `.cell`, yes?

cathysax replied:
Yup! Fixed. (Also updated changeWithoutUndo to use `.link` instead of `.cell`)

cathysax committed:
data-link-id tweaks; new data-reset-prompt

Propagate linked cell (data-link-id) on reset (as for undo); use .link, not .clue (for both).
New option: data-reset-prompt -- optional text string; if set, shown in confirm dialog on reset.

cathysax commented:
In addition to updating to `.link`, I also added a new option: `data-reset-prompt`, which can be used to show a confirm dialog on reset. Let me know if you want me to back that out of this PR and submit separately.

tabascq commented on puzzlejs/puzzle.js:
```diff
@@ -1384,9 +1385,11 @@ function PuzzleEntry(p, index) {
-
+        // TODO shouldn't need a reload after reset
+        if (this.options["data-reset-prompt"])
+            this.commands.querySelector(".puzzle-reset-button").addEventListener("click", e => { if (confirm(this.options["data-reset-prompt"])) {this.prepareToReset(); window.location.reload();} });
```
A few things here:

1. I think these four lines could just boil down to using this handler: `e => { var prompt = this.options["data-reset-prompt"]; if (!prompt || confirm(prompt)) { this.prepareToReset(); window.location.reload(); } }`, so that any changes to the reset logic automatically apply to both cases. That said, see item 2 below which would modify this code somewhat (but the pattern would still work fine IMO).
2. How common do you think it is that the text will change based on the puzzle? I wouldn't think this would happen often, but maybe you have ideas. Just wondering if this should be a bool with standard text (or, if we're feeling fancy, a `data-reset-prompt` bool and a `data-reset-prompt-text` string).
3. If this becomes a bool, I am 100% fine with shifting the default to "true".

cathysax replied:
I wouldn't expect the text to change based on puzzle, but didn't want to assume that others would want the feature or my wording. Making a bool with a generic prompt is fine. What wording do you want? (I'm using "Clear all puzzle content?") Your call on whether the default is true. I'd like that, but how do you feel about changing functionality? Do you also want to include a separate prompt-override string?

cathysax replied:
Actually, I don't think it needs 2 params, especially if you want the default to be showing the prompt. We'd just set `"data-reset-prompt": "Clear all puzzle content?"`, with override options of custom string or "" for no prompt (which would work with your code above since `!prompt` would be true). That functions just like a bool defaulting to true, but without the need to set/check multiple vars.

cathysax replied:
> these four lines could just boil down to using this handler: `e => { var prompt = this.options["data-reset-prompt"]; if (!prompt || confirm(prompt)) { this.prepareToReset(); window.location.reload(); } }`

Changed. I'll wait to upload until we've got the prompt option design resolved.

tabascq replied:
I think I'm now OK with the prompt string being the only option now, especially since `data-reset-prompt = ""` and `data-reset-prompt = "false"` should both succeed in turning off the prompt (since we already convert `"false"` to `false` everywhere). Thanks!

tabascq reviewed (commented):
`.link` changes look good, thanks! As for `data-reset-prompt`: I like having reset confirmation, but have some design thoughts in comments. I'm fine keeping these changes together in one PR, but you can split it if you like.

cathysax committed:
tweak reset prompt option

* tweak reset confirmation option

* example for reset prompt

* tab -> spaces

cathysax commented:
New puzzle.js and reference-options.html uploaded.
